tools/cfs-fuse.c

```diff
@@ -0,0 +1,1225 @@
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stddef.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <sys/mount.h>
+#include <sys/types.h>
+#include <linux/limits.h>
+#include <linux/loop.h>
+#include <linux/mount.h>
+#include <linux/fsverity.h>
+#include <fuse_lowlevel.h>
+#include <sys/mman.h>
+
+#include "libcomposefs/lcfs-erofs.h"
+#include "libcomposefs/lcfs-internal.h"
+
+/* TODO:
+ *  Do we want to user ther negative_timeout=T option?
+ */
+
+#define CFS_ENTRY_TIMEOUT 3600.0
+#define CFS_ATTR_TIMEOUT 3600.0
+
+#define ALIGN_TO(_offset, _align_size)                                         \
+	(((_offset) + _align_size - 1) & ~(_align_size - 1))
+
+/* Note: These only do power of 2 */
+#define __round_mask(x, y) ((__typeof__(x))((y)-1))
+#define round_up(x, y) ((((x)-1) | __round_mask(x, y)) + 1)
+#define round_down(x, y) ((x) & ~__round_mask(x, y))
+
+#include "libcomposefs/erofs_fs_wrapper.h"
+
+const uint8_t *erofs_data;
+size_t erofs_data_size;
+uint64_t erofs_root_nid;
+bool erofs_use_acl;
+const struct erofs_super_block *erofs_super;
+const struct lcfs_erofs_header_s *cfs_header;
+const uint8_t *erofs_metadata;
+const uint8_t *erofs_xattrdata;
+uint64_t erofs_build_time;
+uint32_t erofs_build_time_nsec;
+int basedir_fd;
+
+static void printexit(const char *format, ...)
+{
+	va_list args;
+
+	va_start(args, format);
+	vfprintf(stderr, format, args);
+	va_end(args);
+
+	exit(1);
+}
+
+struct cfs_data {
+	const char *source;
+	const char *basedir;
+	bool noacl;
+};
+
+static const struct fuse_opt cfs_opts[] = {
+	{ "source=%s", offsetof(struct cfs_data, source), 0 },
+	{ "basedir=%s", offsetof(struct cfs_data, basedir), 0 },
+	{ "noacl", offsetof(struct cfs_data, noacl), 1 },
+	FUSE_OPT_END
+};
+
+typedef union {
+	__le16 i_format;
+	struct erofs_inode_compact compact;
+	struct erofs_inode_extended extended;
+} erofs_inode;
+
+static uint64_t cfs_nid_from_ino(fuse_ino_t ino)
+{
+	if (ino == FUSE_ROOT_ID) {
+		return erofs_root_nid;
+	}
+	return ino;
+}
+
+static fuse_ino_t cfs_ino_from_nid(uint64_t nid)
+{
+	if (nid == erofs_root_nid) {
+		return FUSE_ROOT_ID;
+	}
+	return nid;
+}
+
+static const erofs_inode *cfs_get_inode(fuse_ino_t ino)
+{
+	uint64_t nid = cfs_nid_from_ino(ino);
+
+	/* TODO: Add bounds check */
+
+	return (const erofs_inode *)(erofs_metadata + (nid << EROFS_ISLOTBITS));
+}
+
+static uint16_t cfs_inode_version(const erofs_inode *cino)
+{
+	uint16_t i_format = lcfs_u16_from_file(cino->i_format);
+	return (i_format & ((1 << EROFS_I_VERSION_BITS) - 1)) >> EROFS_I_VERSION_BIT;
+}
+
+static bool cfs_inode_is_compact(const erofs_inode *cino)
+{
+	return cfs_inode_version(cino) == 0;
+}
+
+static uint16_t cfs_inode_datalayout(const erofs_inode *cino)
+{
+	uint16_t i_format = lcfs_u16_from_file(cino->i_format);
+	return (i_format & ((1 << EROFS_I_DATALAYOUT_BITS) - 1)) >>
+	       EROFS_I_DATALAYOUT_BIT;
+}
+
+static bool cfs_inode_is_tailpacked(const erofs_inode *cino)
```

Comment:
Kind of here and everywhere...shouldn't we basically `s/cfs_/erofs_`?

Reply:
Not everywhere. This is cfs-fuse, so i think it makes sense to use the cfs_ prefix for it.

However, this particular function, and e few others are very erofs specific, so I guess it makes sense to have them be erofs_...